components/tidb_query_normal_expr/src/builtin_string.rs

```diff
@@ -1001,6 +1002,45 @@ impl ScalarFunc {
             .map(|p| p as i64 + 1)
             .or(Some(0)))
     }
+
+    #[inline]
+    pub fn insert_utf8<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
+        let pos = try_opt!(self.children[1].eval_int(ctx, row));
+        let mut length = try_opt!(self.children[2].eval_int(ctx, row));
+        let newstr = try_opt!(self.children[3].eval_string_and_decode(ctx, row));
+        let input_length = s.chars().count() as i64;
+
+        if pos < 1 || pos > input_length {
+            return Ok(Some(Cow::Owned(s.as_bytes().to_vec())));
+        }
+
+        if length > input_length - pos + 1 || length < 0 {
+            length = input_length - pos + 1
+        }
+
+        let mut inserted = Vec::new();
+        inserted.extend_from_slice(
+            s.chars()
+                .take(pos as usize - 1)
+                .collect::<String>()
+                .into_bytes()
+                .as_vec(),
+        );
```

Comment:
Can we avoid the extra allocation? `char::encode_utf8` may help.

Reply:
I try to keep the style same as [left/right_utf8](https://github.com/tikv/tikv/blob/master/components/tidb_query_normal_expr/src/builtin_string.rs#L273) and [l/rpad_utf8](https://github.com/tikv/tikv/blob/master/components/tidb_query_normal_expr/src/builtin_string.rs#L766), so I think it should be in another pull request to refactor them.

Reply:
@sticnarf So can this PR be accepted? If not, could you please click `insert a suggestion` at here, then I can see how to resolve it. thx~

Reply:
Sorry for the late reply. Actually I think it's different. The implementations you post do not have extra allocations.

Instead of using `extend_from_slice`, I think you can try to `chain` all these iterators and collect once.

Reply:
okay, thank you, I will try to fix it this weekend.